src/ocean/core/Verify.d

```diff
@@ -0,0 +1,82 @@
+/*******************************************************************************
+
+    Utility intended as a replacement for `assert` to check for programming
+    errors and sanity violations in situations when neither removing the check
+    in -release mode nor bringing down the application by throwing an `Error`
+    is acceptable.
+
+    This module must have as few import dependencies as possible so that it can
+    be used in place of `assert` freely without introducing cyclic imports.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.core.Verify;
+
+import ocean.meta.types.Qualifiers : istring;
+
+/*******************************************************************************
+
+    Verifies that certain condition is met.
+
+    Params:
+        ok = boolean condition to check
+        msg = optional exception message
+
+    Throws:
+        SanityException if `ok` condition is `false`.
+
+*******************************************************************************/
+
+public void verify ( bool ok, lazy istring msg = "",
+    istring file = __FILE__, int line = __LINE__ )
+{
+    static SanityException exc;
+
+    if (exc is null)
+        exc = new SanityException("");
+
+    if (!ok)
+    {
+        exc.file = file;
+        exc.line = line;
+        exc.msg = msg;
+
+        throw exc;
+    }
+}
+
+///
+unittest
+{
+    try
+    {
+        verify(false);
+    }
+    catch (SanityException e) { }
+
+    verify(true);
+}
+
+/*******************************************************************************
+
+    Indicates some internal sanity violation in the app, essentially a less
+    fatal version of `AssertError`.
+
+*******************************************************************************/
+
+public class SanityException : Exception
```

Comment:
There was no specific discussion, I picked the name because I was already using this name in other places to name things with such purpose. I believe it is a more generic and thus correct term than `ProgrammingException` because cases when we would want to use it may also detect architecture issues or serious violations from trusted input source (like DHT).